_posts/2021-04-28-#17331.md

```diff
@@ -0,0 +1,145 @@
+
+    3. The wallet's UTXO pool is retrieved via `AvailableCoins()`.
+
+    4. Using the above fixed values, `SelectCoins()` is used to pick an input
+       set to fund the transaction. If the user manually selected some UTXOs by
+       means of the `CoinControl` feature, these UTXOs are used first. The
+       remaining UTXOs are grouped into `OutputGroup`s. The grouped UTXOs are
+       then used in one or multiple rounds of `SelectCoinsMinConf()` which
+       use increasingly permissive `CoinEligibilityFilter`s to preselect
+       `OutputGroup`s until an input set is found. As described above,
+       `SelectCoinsMinConf()` tries BnB first, then falls back to knapsack if no
+       solution was found.
+    5. Determine whether to create a change output, finish assembling the
+       transaction, and sign the transaction.
+
+- When a UTXO is selected as an input, the size of the transaction is increased
+  which in turn increases the necessary fees to maintain the targeted feerate.
+  We can preempt the cost added by the input by considering each UTXO at its
+  _effective value_, its value reduced by the UTXO's input cost at the given
+  feerate.
+
+- Using _effective values_ in coin selection means that we can freely add and
+  remove inputs until we have have raised enough funds to pay for the
+  transaction's fixed costs: the payment amounts, the transaction header and the
+  output data. Whatever final input set we settle for, the increased transaction
+  size and input costs have already been accounted for.
+
+- The _knapsack solver_ sorts the available UTXOs by effective value, and then
+  combines candidate input sets by randomly selecting or skipping inputs until
+  it has sufficient funds for the transaction. When it finds a viable solution,
+  it keeps the solution if it produces a smaller change than the prior stored
+  solution. Then, the algorithm deselects the last input and continues
+  traversing the UTXO list as described, trying to converge on the smallest
+  change output above `MIN_CHANGE`. This random process is repeated up to 1,000
+  times.
+
+- The _Branch and Bound solver (BnB)_ deterministically searches the complete
+  combination space to find an input set that will avoid the creation of a
+  change output. It performs a depth-first search on a binary tree where each
+  branch represents the inclusion or exclusion of a UTXO, exploring inclusion
+  branches first, and backtracking whenever a subtree cannot yield a solution.
+  It returns the first discovered input set that is an "exact match" for the
+  funding requirement of the transaction. The BnB algorithm is not guaranteed to
+  find a solution even when there are sufficient funds since an exact match may
+  not exist.
+
+- An
+  [OutputGroup](https://github.com/bitcoin-core-review-club/bitcoin/blob/4ac1adda9914d845aaea5804af4801ffec53c701/src/wallet/coinselection.h#L72)
+  collects all UTXOs that were sent to the same public key.
```

Comment:
I think this can either be deleted or moved up to the first bullet

Reply:
Moved up.